### ISSUE TITLE
if statement local variable in else clause

### ISSUE THREAD

**Issue by dc0d** [golang/lint#301]
`golint` shows a warning for an `if` statement local variable, used in it's `else` clause, when `if` returns, such as:

```go
if v, err := s.marshal(doc); err != nil {
	return err
} else { // should not show warning
	value = v
}
```

**Comment by curio77** [comment]
I was just about to report that too, so +1!

**Comment by dc0d** [comment]
At least there should be this option to justify a certain usage with a comment (such as in `_ "some-package"` imports).

This way we can help the linter and ourselves. Otherwise we have to have lots of warnings when they are not really warnings and that makes the linter messages meaningless.

**Comment by TomasB** [comment]
@dc0d +1

@dsymonds:
> I wrote the original check, so I know what the intent was

Did you account though that the warning will make  'If with a short statement' not usable in some very common scenarios?

We can't use 'if with short statement' where function returns (value, error): 
```
if v, err := f(); err {
   log.Error(err)
} else {
   mem = v
}
```

Instead, we have to write:
```
v, err := f();
if err {
   log.Error(err)
}
mem = v
```
which isn't as clean, since:
- we expanded scope for v, which might lead to errors.
- we can't write consequent blocks of similar code using same temp var names 

if course, we could write this with additional {}:
```
{
    v, err := f();
    if err {
       log.Error(err)
    }
    mem = v
}
```
but where is fun in that?

Given the choice, we (devs) would rather take if with short statement over that warning.
Number of opened tickets for that erroneous warning supports this statement.

**Comment by l0k1verloren** [comment]
This warning is an *error*, triggering on an else block using a variable declared inside the if block scope. An error. I wouldn't mind so much if I could filter these warnings out entirely but then I might also miss opportunities to simplify the code as well.

**Comment by dc0d** [comment]
@mvdan Thank you for the update.

Just for the record: [golangci-lint](https://github.com/golangci/golangci-lint) is a very helpful and thorough alternative.
